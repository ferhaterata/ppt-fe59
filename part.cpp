//  ----------------------------------------------------------------------------
//  Implementation file for the Part class.                             part.cpp
//  Created by Ferhat Erata <ferhat.erata@yale.edu> on December 12, 2019.
//  Copyright (c) 2019 Yale University. All rights reserved.
// -----------------------------------------------------------------------------

#include "part.hpp"
#include "exception.hpp"
#include <iomanip>

// -----------------------------------------------------------------------------
// Part class should be an expert at extracting the part record from the file
std::istream& Part::read(std::istream& in) {
    in >> name >> quantity >> price;
    if (in.good() and quantity < 0) {
        std::stringstream ss;
        ss << "(" << *this << "):"
           << " quantity cannot be negative!";
        throw IOerror(ss.str());
    }
    return in;
}

// -----------------------------------------------------------------------------
// Part is an expert at serializing its fields. The fields comprising a part
// should be separated by a single space. The price should be printed as a
// decimal number with two digits after the decimal point
std::ostream& Part::write(std::ostream& out) const {
    return out << std::fixed << name << " " << quantity << " "
               << std::setprecision(2) << price;
}
