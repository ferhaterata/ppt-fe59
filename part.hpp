//  ----------------------------------------------------------------------------
//  Header file for the Part class.                                     part.hpp
//  Created by Ferhat Erata <ferhat.erata@yale.edu> on December 12, 2019.
//  Copyright (c) 2019 Yale University. All rights reserved.
// -----------------------------------------------------------------------------

#ifndef REPAIR_PART_HPP
#define REPAIR_PART_HPP

#include <iostream>
#include <string>

// -----------------------------------------------------------------------------
// A company sells many different kinds of parts
struct Part {
    std::string name;   // part's name
    int quantity = 0;   // the quantity on hand
    double price = 0.0; // the selling price

    std::istream& read(std::istream& in);
    std::ostream& write(std::ostream& out) const;
};

// deserialize a part from input stream
inline std::istream& operator>>(std::istream& in, Part& part) {
    return part.read(in);
}

// serialize a part to output stream
inline std::ostream& operator<<(std::ostream& out, const Part& part) {
    return part.write(out);
}

#endif // REPAIR_PART_HPP
