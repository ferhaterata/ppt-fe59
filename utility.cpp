//  ----------------------------------------------------------------------------
//  Implementation file for the utility classes.                     utility.cpp
//  Created by Ferhat Erata <ferhat.erata@yale.edu> on December 12, 2019.
//  Copyright (c) 2019 Yale University. All rights reserved.
// -----------------------------------------------------------------------------

#include "utility.hpp"
#include <ctime>
#include <iomanip>
#include <iostream>

void utility::banner() {
    std::time_t t = std::time(nullptr);
    std::tm tm = *std::localtime(&t);
    std::cout << "\n" << separator();
    std::cout << "\t"
              << "Ferhat Erata"
              << "\n\t"
              << "CPSC 427/527"
              << "\n\t" << std::put_time(&tm, "%c %Z");
    std::cout << "\n" << separator() << std::endl;
}

void utility::bye() {
    std::cout << "\n" << separator() << "Normal termination.\n";
}

std::string utility::separator() {
    return "---------------------------------------------------------------\n";
}

std::string utility::print_stream_state(const std::ios& stream) {
    std::stringstream ss;
    ss << " good()=" << stream.good();
    ss << " eof()=" << stream.eof();
    ss << " fail()=" << stream.fail();
    ss << " bad()=" << stream.bad() << std::endl;
    return ss.str();
}