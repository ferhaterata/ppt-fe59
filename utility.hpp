//  ----------------------------------------------------------------------------
//  Header file for the utility classes.                             utility.hpp
//  Created by Ferhat Erata <ferhat.erata@yale.edu> on December 12, 2019.
//  Copyright (c) 2019 Yale University. All rights reserved.
// -----------------------------------------------------------------------------

#ifndef UTILITY_HPP
#define UTILITY_HPP

#ifdef DEBUG

#define print(a) std::cout << a << std::endl

#else
#define DEBUG 0
#define print(a)
#endif

#include <string>

namespace utility {

// -----------------------------------------------------------------------------
// Prints a neat header on the output.
void banner();

// -----------------------------------------------------------------------------
//  Prints termination message.
void bye();
void ferhat();
// -----------------------------------------------------------------------------
//  Generates a separator string
std::string separator();

// -----------------------------------------------------------------------------
// Prints stream state for debugging purposes.
std::string print_stream_state(const std::ios& stream);

} // namespace utility

#endif // UTILITY_HPP
