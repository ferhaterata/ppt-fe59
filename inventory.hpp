//  ----------------------------------------------------------------------------
//  Header file for the Inventory class.                           inventory.hpp
//  Created by Ferhat Erata <ferhat.erata@yale.edu> on December 12, 2019.
//  Copyright (c) 2019 Yale University. All rights reserved.
// -----------------------------------------------------------------------------

#ifndef REPAIR_INVENTORY_HPP
#define REPAIR_INVENTORY_HPP

#include "part.hpp"
#include <vector>

// -----------------------------------------------------------------------------
// Contains a list of parts, represented by a vector, as well as functions for
// reading and writing inventory files and for repairing the parts list.
class Inventory {
  private:
    std::vector<Part> parts; // initializes an empty vector of parts.

  public:
    Inventory() = default;

    std::ostream& write(std::ostream& out) const;
    std::istream& read(std::istream& in); // extracts parts and fills the vector

    Inventory repair(); // repair the parts and returns a new inventory object
};

// deserialize an inventory from input stream
inline std::istream& operator>>(std::istream& in, Inventory& inventory) {
    return inventory.read(in);
}

// serialize an inventory to output stream
inline std::ostream& operator<<(std::ostream& out, const Inventory& inventory) {
    return inventory.write(out);
}

// -----------------------------------------------------------------------------
// a functor that orders two Part objects alphabetically based on a string
// comparison of their names.
class CompareByName {
  public:
    bool operator()(const Part& lhs, const Part& rhs) {
        return (lhs.name < rhs.name);
    }
};

#endif // REPAIR_INVENTORY_HPP
