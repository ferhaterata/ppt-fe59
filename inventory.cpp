//  ----------------------------------------------------------------------------
//  Implementation file for the Inventory class.                   inventory.cpp
//  Created by Ferhat Erata <ferhat.erata@yale.edu> on December 12, 2019.
//  Copyright (c) 2019 Yale University. All rights reserved.
// -----------------------------------------------------------------------------

#include "inventory.hpp"
#include "utility.hpp"
#include <algorithm>

// -----------------------------------------------------------------------------
// Inventory class should be an expert at extracting the all records
std::istream& Inventory::read(std::istream& in) {
    do {
        Part part;
        if (in >> part) {          // !in.fail()
            parts.push_back(part); // got good data
            print(part);
        }
        if (in.fail()) {
            print(utility::print_stream_state(in));
            print("(" << part << ")\n");
        }
    } while (in.good());
    return in;
}

// -----------------------------------------------------------------------------
// Inventory class is an expert at serializing all the part data.
// ensures: The output file should have each part on a separate line.
std::ostream& Inventory::write(std::ostream& out) const {
    for (const auto& part : parts) {
        out << part << "\n";
    }
    return out;
}

// -----------------------------------------------------------------------------
// ensures: If the original file contains multiple entries for the same part,
// that partâ€™s quantity on hand in the new file should be the sum over those
// entries of the quantities on hand in the old file. The price of that part in
// the new file should be the maximum price of that part in the old file. There
// should be only one entry for each part. Records should be sorted
// alphabetically by part name
Inventory Inventory::repair() {
    // sort alphabetically based on a string comparison of their names.
    std::sort(parts.begin(), parts.end(), CompareByName());
    print(utility::separator());
    print(*this);
    Inventory repaired;         // creates a new Inventory object to return
    auto first = parts.begin(); // iterator
    auto last = parts.end();    // mark last
    while (first != last) {     // it makes a pass over the old inventory list
        auto head = first++;
        double max = head->price;
        // merges together each block of entries for the same part,
        while (first != last and head->name == first->name) {
            head->quantity += first->quantity;
            if (first->price > max) {
                max = first->price;
            }
            ++first;
        }
        // The price of that part in the new file should be the maximum price of
        // that part in the old file.
        head->price = max;
        // and pushes the resulting parts onto the back of the new inventory
        // object
        repaired.parts.push_back(*head);
    }
    print(utility::separator());
    print(repaired);
    return repaired;
}
