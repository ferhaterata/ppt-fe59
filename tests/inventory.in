AWF  5 45
MJF  3   98.65
FDR  5   75.5
GLD  9  5
RBW  3   69.001
GBS  1   92.74
PLK  16  37.88
AWF  24  56
PDB  71  86
JBK  1  24
GLD  89  53
AWF  68  100.006
HST  79  06.999
ABC  82  05.99
AEF  89  1405.003
ALA  105 1405
PLK  5  37
ALA  105 22
GLD  89  53