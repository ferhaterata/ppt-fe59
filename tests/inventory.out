ABC 82 5.99
AEF 89 1405.00
ALA 210 1405.00
AWF 97 100.01
FDR 5 75.50
GBS 1 92.74
GLD 187 53.00
HST 79 7.00
JBK 1 24.00
MJF 3 98.65
PDB 71 86.00
PLK 21 37.88
RBW 3 69.00
